[#77] - Implement endpoints for sharing posts

Dj5822 opened the pull request:
# Description:
Added share and unshare, which are used to share and unshare a post.
Added tests to make sure that these endpoints are working.
Also fixed an error in the migration where the likedPosts and sharedPosts were not getting dropped correctly.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](../CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project
- [ ] My code has been commented
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

Dj5822 committed:
Updated the database

Created two new tables: likedpost and sharedpost to store the associations between posts and users.

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into posts-dto

Dj5822 committed:
Added dto for posts

Uses dto for POST and GET method.

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into posts-dto

Dj5822 committed:
Updated error codes

Updated the error codes and message such that 500 is sent instead of 400 if the database modification fail.

Dj5822 committed:
Updated usersLiked and usersShared

Changed usersLiked and usersShared in the posts dto such that they return the actual number of likes and shares.

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into posts-dto

Dj5822 committed:
Converted datatypes to integers

Dj5822 committed:
Fixed typo

Dj5822 committed:
Modified tests to check the contents of the dto

Checks certain values such as the content, author's username and email, parent, children, usersLiked, and usersShared.

Dj5822 committed:
Improved tests further

Now checks whether the id exists and whether the timestamp exists. Doesn't check whether these values are valid though.

Dj5822 committed:
Added endpoints for share and unshare post

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into shared-posts

Dj5822 committed:
Merge branch 'SE701Team2:main' into shared-posts

Dj5822 committed:
Fixed bug

Fixed bug where the table names from dropping sharedPosts and likedPosts was incorrect.

Dj5822 committed:
Removed console messages

Dj5822 committed:
Created outline for tests

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into shared-posts

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into shared-posts

Dj5822 committed:
Implemented tests for sharing posts

Dj5822 committed:
Added tests for unsharing

mdar365 commented on backend/__tests__/posts.js:
```diff
@@ -402,4 +402,106 @@ describe('Posts', () => {
+                const authToken = Authentication.generateAuthToken(user1)
+
+                const response = await request(server)
+                    .post(`/api/posts/1/share`)
```
Id of the post shouldnt be hardcoded because id is automatically generated so it could be different for everyone. We should be creating a new post in the test in case our DBs are cleared

Dj5822 replied:
I got a little lazy and cut corners there. I think I fixed it. There are still some hard coded values for the tests involving unliking, but they shouldn't be a problem.

mdar365 reviewed (changes requested):
We just need to change the tests so the id isn't hard-coded. I also suggest creating a new post in each of those tests and using its ID as our DBs have different data

Dj5822 committed:
Removed hard coded values from tests

mdar365 commented on backend/__tests__/posts.js:
```diff
@@ -402,4 +402,124 @@ describe('Posts', () => {
+
+                const authToken = Authentication.generateAuthToken(user1)
+
+                await request(server)
+                    .post(`/api/posts/2/share`)
+                    .set('Authorization', `Bearer ${authToken}`)
```
Here you're sharing a post that you haven't created, wouldnt that cause an error if we don't have that post?

Also, would it be possible to only send requests to the endpoints that we're testing? The rest can be added directly into the db using models.create(), so that changes to the POST /share endpoint does not cause it to fail

Dj5822 replied:
Okay I fixed it.

mdar365 commented on backend/__tests__/posts.js:
```diff
@@ -402,4 +402,124 @@ describe('Posts', () => {
+                    .send({
+                        text_content: 'what is the meaning of life?',
+                        parent: null,
+                    })
```
Same here, we're testing the /share endpoint so we shouldnt bring another endpoint into the test if we don't need to, could you fix it for the other tests as well?

Dj5822 committed:
Fixed tests

Dj5822 committed:
Updated tests

Removed all instances where endpoint in a test that doesn't involve that endpoint

Dj5822 committed:
Merge branch 'main' of https://github.com/Dj5822/Updog into shared-posts

mdar365 commented on backend/__tests__/posts.js:
```diff
@@ -402,4 +400,125 @@ describe('Posts', () => {
+    describe('DELETE /posts/:id/share', () => {
+        describe('when not authenticated', () => {
+            it('should return response code of 400', async () => {
+                const response = await request(server).post(
+                    '/api/posts/1/share'
+                )
```
Should this one be delete?

Dj5822 replied:
That is correct...

Dj5822 committed:
Fixed test case where post was used instead of delete
